pkg/agent/runner.go

```diff
@@ -365,17 +429,28 @@ func (r *Runner) getMetricsLoop(
 	}
 
 	for {
-		metrics, err := r.doMetricsRequest(ctx, logger, timeout)
-		if err != nil {
-			logger.Error("Error making metrics request", zap.Error(err))
-			goto next
-		} else if metrics == nil {
-			goto next
-		}
+		if !mgr.isActive() {
+			if lastActive {
+				logger.Info(fmt.Sprintf("VM is no longer active for %s metrics requests", mgr.kind))
+			}
+			lastActive = false
+		} else {
+			if !lastActive {
+				logger.Info(fmt.Sprintf("VM is now active for %s metrics requests", mgr.kind))
+			}
+			lastActive = true
 
-		newMetrics(*metrics, func() {
-			logger.Info("Updated metrics", zap.Any("metrics", *metrics))
-		})
+			metrics := M(new(V)) // construct M as a pointer to the zero value. This is what V is for.
+			err := doMetricsRequest(r, ctx, logger, metrics, config)
+			if err != nil {
+				logger.Error("Error making metrics request", zap.Error(err))
+				goto next
+			}
+
+			mgr.newMetrics(metrics, func() {
+				logger.Info("Updated metrics", zap.Any("metrics", metrics))
```

Comment:
I know this is a prior change, but why this callback exists, and why is named `withLock`? Wouldn't just adding this `logger.Info` wherever `withLock` is called serve the same purpose, only simpler?

Reply:
I think I did it at the time because it felt like a cleaner separation of concerns -- the caller is doing the logging anyways, so the executor stuff only needs to be concerned with doing the updating. 

Looking at it now, I'm not sure either way, and hesitant to change it. Happy to discuss more though.

Reply:
> it's easier if that doesn't also have logging

It doesn't have logging, but if have to pass the callback, so not much profit :upside_down_face: 

> I'm not sure either way, and hesitant to change it.

No problem, that's was before the current change.